### PR TITLE
Add USB support for STM32G0 series

### PR DESCRIPTION
**Pull request by ycsin**
Add USB support for STM32G0 series.
SVD file with USB added for STM32G0B1 series can be downloaded [here](https://gist.github.com/ycsin/8293fd5eeccdbdc7c939f6e138e31625).

Signed-off-by: Yong Cong Sin <yongcong.sin@gmail.com>

### CODE REVIEW

**Comment by ycsin** [comment]
Tested the `cdc_acm` and `shell/shell_module` sample with my stm32g0b1re custom board.

**Comment by ycsin** [comment]
Ping @erwango @ABOSTM @FRASTM @str4t0m

**Comment by ycsin** [comment]
I noticed that it would take a second try  for the `shell_module` sample to work after unplugging the USB:

--- Sample works
--- Unplug USB
--- Plug in
--- Sample doesn't work
--- Unplug USB
--- Plug in
--- Sample works
--- Unplug USB
--- Plug in
--- Sample doesn't work
--- Unplug USB
--- Plug in
... repeat

**Review by erwango — CHANGES_REQUESTED** [review]
Thanks for this work!
This looks good aside from one comment on pure driver implementation.

I'd have 2 more requests before pushing this work:
- add USB node on stm32g0b0.dtsi
- add support on one in tree board (nucleo_g0b1re), this would allow driver compilation and test in CI. Is that possible for you ?

File: drivers/usb/device/usb_dc_stm32.c [hunk 1]
```diff
@@ -344,8 +344,12 @@ static int usb_dc_stm32_init(void)
 	HAL_StatusTypeDef status;
 	unsigned int i;
 
+#if defined(USB) || defined(USB_DRD_FS)
 #ifdef USB
```
[the remark this answers is not in the document]
Comment (ycsin) [reply drivers/usb/device/usb_dc_stm32.c:348]: @erwango I don't understand, the `#ifdef USB` here can't be changed to `#elif USB`, this is a nested #if

[the remark this answers is not in the document]
Comment (erwango) [reply drivers/usb/device/usb_dc_stm32.c:348]: ok, got it. This part of code is not quite satisfying and would need to be reviewed, but for now this change is ok.

File: drivers/usb/device/usb_dc_stm32.c [hunk 2]
```diff
@@ -75,7 +75,7 @@ static const struct soc_gpio_pinctrl usb_pinctrl[] =
  * WARNING: Don't mix USB defined in STM32Cube HAL and CONFIG_USB from Zephyr
  * Kconfig system.
  */
-#ifdef USB
+#if defined(USB) || defined(USB_DRD_FS)
```
Comment (loicpoulain) [inline drivers/usb/device/usb_dc_stm32.c:78]: Could you please update the comment above, explaining what USB_DRD_FS represents?

Comment (ycsin) [reply drivers/usb/device/usb_dc_stm32.c:78]: "USB dual role device full speed" IIRC, that's defined by ST here:
https://github.com/zephyrproject-rtos/hal_stm32/blob/main/stm32cube/stm32g0xx/soc/stm32g0b0xx.h#L701

**Comment by ycsin** [comment]
> Thanks for this work!
> This looks good aside from one comment on pure driver implementation.
> 
> I'd have 2 more requests before pushing this work:
> 
> * add USB node on stm32g0b0.dtsi
> * add support on one in tree board (nucleo_g0b1re), this would allow driver compilation and test in CI. Is that possible for you ?

I'll do that sometime later.

**Comment by ycsin** [comment]
> Thanks for this work!
> This looks good aside from one comment on pure driver implementation.
> 
> I'd have 2 more requests before pushing this work:
> 
> * add USB node on stm32g0b0.dtsi
> * add support on one in tree board (nucleo_g0b1re), this would allow driver compilation and test in CI. Is that possible for you ?

- [X] add USB node on stm32g0b0.dtsi
- [X] add support on one in tree board (nucleo_g0b1re), this would allow driver compilation and test in CI. Is that possible for you

File: drivers/usb/device/usb_dc_stm32.c (hunk 2 above)
Comment (loicpoulain) [reply drivers/usb/device/usb_dc_stm32.c:78]: yes but update the comment above since it only mention USB and USB_OTG_FS types.

Comment (ycsin) [reply drivers/usb/device/usb_dc_stm32.c:78]: Got it

**Review by loicpoulain — APPROVED** [review]

**Review by erwango — APPROVED** [review]

**Pull request merged**